contracts/ScopeGuard.sol

```diff
@@ -4,19 +4,51 @@ pragma solidity ^0.8.6;
 import "@gnosis.pm/zodiac/contracts/guard/BaseGuard.sol";
 import "@gnosis.pm/zodiac/contracts/factory/FactoryFriendly.sol";
 
+enum Clearance {
+    None,
+    Target,
+    Function
+}
+
+enum ExecutionOptions {
+    None,
+    Send,
+    DelegateCall,
+    Both
+}
+
 contract ScopeGuard is FactoryFriendly, BaseGuard {
-    event SetTargetAllowed(address target, bool allowed);
-    event SetTargetScoped(address target, bool scoped);
-    event SetFallbackAllowedOnTarget(address target, bool allowed);
-    event SetValueAllowedOnTarget(address target, bool allowed);
-    event SetDelegateCallAllowedOnTarget(address target, bool allowed);
-    event SetFunctionAllowedOnTarget(
-        address target,
-        bytes4 functionSig,
-        bool allowed
-    );
+    event AllowTarget(address target);
+    event RevokeTarget(address target);
+    event ScopeTarget(address target);
+
+    event SetExecutionOptions(address target, ExecutionOptions options);
+
+    event ScopeAllowFunction(address target, bytes4 functionSig);
+    event ScopeRevokeFunction(address target, bytes4 functionSig);
+
+    event ScopeAllowFallback(address target, bytes4 functionSig);
+    event ScopeRevokeFallback(address target, bytes4 functionSig);
+
     event ScopeGuardSetup(address indexed initiator, address indexed owner);
 
+    /// Function signature too short
+    error FunctionSignatureTooShort();
+
+    /// Role not allowed to send to target address
+    error SendNotAllowed();
+
+    /// Role not allowed to delegate call to target address
+    error DelegateCallNotAllowed();
+
+    /// Role not allowed to call target address
+    error TargetAddressNotAllowed();
+
+    /// Role not allowed to call this function on target address
+    error FunctionNotAllowed();
+
+    bytes4 internal constant FALLBACK_FUNCTION_SIG = 0x00000000;
```

Comment:
How to disambiguate between the fallback function and a function with the signature `0x00000000`?

Reply:
I thought they were the same? I never thought about this, but this is true, technically there could be some signature out there that yields 0x00000000

I chatted about this Adam, and he agreed that in roles we would handle the fallback function case via this function signature, the bytes4(0)